GTK 4 support

TheTumultuousUnicornOfDarkness opened the pull request:
Hello,

This PR add GTK 4 support in linuxdeploy-plugin-gtk. According to [this message](https://gitlab.gnome.org/GNOME/gtk/-/issues/1181#note_255975), GTK+ 4.0 do not support IM modules.
I removed explicit mentions to GTK 2 and GTK 3 to cover GTK 4.

I added Dockerfiles to test the plugin with a GTK3 and a GTK4 application (the widget factory) on Debian, Fedora, openSUSE and Ubuntu.
There is an minor issue on the GitHub Runner with the shared volume (`/AppImage` inside containers), but I do not have this problem locally. It does not matter, because the volume store built AppImages and we do not really need them in fact.

Please note this PR is based on #19.

TheTumultuousUnicornOfDarkness committed:
Check if find is installed

TheTumultuousUnicornOfDarkness committed:
Add GTK4 support

Close #17

TheTumultuousUnicornOfDarkness committed:
Update README

TheTumultuousUnicornOfDarkness committed:
Add Dockerfiles for tests

TheTumultuousUnicornOfDarkness committed:
Add CI job to build containers

TheTumultuousUnicornOfDarkness committed:
CI: set DEPLOY_GTK3 variable

TheAssassin commented:
Are you saying that Gtk+2 is no longer supported?

TheAssassin commented on linuxdeploy-plugin-gtk.sh:
```diff
@@ -101,6 +144,23 @@ if [ -z "$LINUXDEPLOY" ]; then
+fi
+
+if variable_is_true "$DEPLOY_GTK3" && variable_is_true "$DEPLOY_GTK4"; then
+    echo -e "$0: can not deploy multiple GTK versions.\nSet only DEPLOY_GTK3 or DEPLOY_GTK4 environment variable."
```
That message is a bit misleading. How. Also I'd rather split it up to two echo calls to reduce the line length. Given that they describe major versions of Gtk+, how about `DEPLOY_GTK_VERSION={3, 4}`?

```
[...] Please set DEPLOY_GTK_VERSION to {3, 4}.

TheTumultuousUnicornOfDarkness replied:
Fair enough. Fixed in 5c8f98e, thanks for the feedback.

TheTumultuousUnicornOfDarkness commented:
Well, before my change, I found Gtk+2 was not really supported, because the plugin does not copy Gtk+2 modules. Maybe it was working partially, but I guess if it works, it is because it uses system libraries.
Gtk+2 is EOL, I am not sure if it is worth it to add full Gtk+2 support in 2021.

TheAssassin commented:
Doesn't have to be full support, let's call it "best effort". There's still quite a bunch of Gtk+2 apps out there, e.g., anything written for DEs like XFCE or MATE. I just don't want to break existing deployment setups.

TheAssassin commented:
I see, you haven't actually removed any functionality. Well, that's fine then, I guess.

TheTumultuousUnicornOfDarkness commented:
Let me add a new commit to allow people to be able to still use this plugin with Gtk+2 applications.

TheTumultuousUnicornOfDarkness committed:
Re-add Gtk+2 support

TheTumultuousUnicornOfDarkness commented:
With c5af397ba3f9f8a9234af689ba82edb5978ec174, I restored the old behavior for Gtk+2 apps.
I think we are good now.

TheTumultuousUnicornOfDarkness commented:
@TheAssassin do I need to change something else?

TheAssassin reviewed (commented):
Somehow this review got lost >6 weeks ago. I just found it on GitHub. Sorry for the delay.

TheTumultuousUnicornOfDarkness committed:
Store GTK version in DEPLOY_GTK_VERSION

TheTumultuousUnicornOfDarkness committed:
CI: fix errors

find: ‘/tmp/AppDir/usr/bin’: No such file or directory

TheAssassin commented on .github/workflows/containers.yml:
```diff
@@ -0,0 +1,37 @@
+
+jobs:
+  containers:
+    name: Buildah
```
When using a strategy, it makes sense to a) first list the strategy, then b) include the values from the matrix in the name.

TheTumultuousUnicornOfDarkness replied:
Thanks for the tip, fixed in f297691.

TheAssassin commented on .github/workflows/containers.yml:
```diff
@@ -0,0 +1,37 @@
+      - name: Build container
+        run: |
+          mkdir -pv "${ARTIFACT_DIR}"
+          buildah bud \
```
What's the difference to the more standard Docker command line? I haven't heard of this tool until 5 minutes ago.

TheTumultuousUnicornOfDarkness replied:
Buildah build rootless containers easily.
~Docker on GitHub Actions requires [QEMU to work](https://github.com/marketplace/actions/build-and-push-docker-images#git-context), that is overkill.~
EDIT: Ah, you mean like [this](https://docs.github.com/en/actions/creating-actions/creating-a-docker-container-action#creating-an-action-metadata-file)? I am not familiar with that.

TheTumultuousUnicornOfDarkness replied:
I remember the "why" now.
Buildah allows to mount volume during the build, while `docker build` does not.
The idea is to mount a volume under `/AppImage` during the container build, so all built AppImages are available in this directory.
Then, we can create artifacts from AppImages easily (because `docker cp` only works with running containers, not with images).

TheAssassin replied:
Why do you need to build the AppImages during the container build, though? The typical workflow is to build a Docker image with the dependencies, then run an instance that executes a build script.

Anyway, I don't really care, let's merge this!

TheAssassin commented on .github/workflows/containers.yml:
```diff
@@ -0,0 +1,37 @@
+
+      - uses: actions/upload-artifact@v2
+        with:
+          name: "${LINUX_DISTRO} with ${GTK_VERSION}"
```
I'm concerned that this is going to cause quota issues.

TheAssassin replied:
Also, I don't understand why GitHub's own image registry is not used.

TheTumultuousUnicornOfDarkness replied:
The idea was to store built AppImages as artifact, not the container image itself.
It is more for troubleshooting so I can delete this step if needed.

TheAssassin commented on .github/workflows/main.yml:
```diff
@@ -7,13 +7,17 @@ jobs:
       run: |
         echo "REPODIR=$GITHUB_WORKSPACE" >> $GITHUB_ENV
         echo "APPDIR=/tmp/AppDir"        >> $GITHUB_ENV
+        echo "DEPLOY_GTK_VERSION=3"      >> $GITHUB_ENV
```
Why not use the `env:` key only? This is somewhat redundant.

TheTumultuousUnicornOfDarkness replied:
You are right. Fixed in 2d9be7b8376b4a99a737092aa841ae6a5def02d7.

TheAssassin commented on containers/gtk3/Dockerfile.fedora:
```diff
@@ -0,0 +1,22 @@
+FROM docker.io/fedora:33 AS build-stage
```
Why do you specify the `docker.io` (default) registry manually? I'm curious whether there is any kind of advantage in doing so.
Of course, it's fine, functionality wise.

TheTumultuousUnicornOfDarkness replied:
This is required with `buildah`, because the `docker.io` registry is not implicitly set by default.

TheAssassin commented on linuxdeploy-plugin-gtk.sh:
```diff
@@ -67,6 +85,7 @@ search_tool() {
 }
 
 APPDIR=
+DEPLOY_GTK_VERSION="${DEPLOY_GTK_VERSION:-0}"
```
Please add a comment that states that `0` indicates "not found".

Code style wise, I think you should move it above the `APPDIR` variable. The latter is set in the following while loop, and logically belongs to that block.

TheTumultuousUnicornOfDarkness replied:
Ok, fixed in cd303e99a97a61d599366dd71c98fdc334d4d4bc.

TheAssassin commented on linuxdeploy-plugin-gtk.sh:
```diff
@@ -108,6 +127,11 @@ else
 fi
 
+if ! command -v find &>/dev/null && ! type find &>/dev/null; then
+    echo -e "$0: find not found.\nInstall findutils then re-run the plugin."
```
:+1:

TheTumultuousUnicornOfDarkness committed:
CI: use 'env' key for environment variables

TheTumultuousUnicornOfDarkness committed:
Add a comment about DEPLOY_GTK_VERSION

TheTumultuousUnicornOfDarkness committed:
CI: include the values from the matrix in the name

TheTumultuousUnicornOfDarkness committed:
CI: fix volume options for buildah

It looks like the 'U' flag does not exit anymore
http://manpages.ubuntu.com/manpages/groovy/man1/buildah-bud.1.html

TheTumultuousUnicornOfDarkness committed:
CI: fix upload-artifact

qarmin commented:
Still some features are missing?
I tested this with binary - [czkawka_gui.zip](https://github.com/linuxdeploy/linuxdeploy-plugin-gtk/files/7637357/czkawka_gui.zip) and it works fine
